sockeye/scoring.py

```diff
@@ -0,0 +1,266 @@
+
+            # decoder
+            # target_decoded: (batch-size, target_len, decoder_depth)
+            target_decoded = self.decoder.decode_sequence(source_encoded, source_encoded_length, source_encoded_seq_len,
+                                                          target_embed, target_embed_length, target_embed_seq_len)
+
+            # output layer
+            # logits: (batch_size * target_seq_len, target_vocab_size)
+            logits = self.output_layer(mx.sym.reshape(data=target_decoded, shape=(-3, 0)))
+            # logits after reshape: (batch_size, target_seq_len, target_vocab_size)
+            logits = mx.sym.reshape(data=logits, shape=(-4, -1, target_embed_seq_len, 0))
+
+            if self.softmax_temperature is not None:
+                logits = logits / self.softmax_temperature
+
+            # Compute the softmax along the final dimension.
+            # target_dists: (batch_size, target_seq_len, target_vocab_size)
+            target_dists = mx.sym.softmax(data=logits, axis=2, name=C.SOFTMAX_NAME)
+
+            # Select the label probability, then take their logs.
+            # probs and scores: (batch_size, target_seq_len)
+            probs = mx.sym.pick(target_dists, labels)
+            scores = mx.sym.log(probs)
+            if self.score_type == C.SCORING_TYPE_NEGLOGPROB:
+                scores = -1 * scores
+
+            # Sum, then apply length penalty. The call to `mx.sym.where` masks out invalid values from scores.
+            # zeros and sums: (batch_size,)
+            zeros = mx.sym.zeros_like(scores)
+            sums = mx.sym.sum(mx.sym.where(labels != 0, scores, zeros), axis=1) / (self.length_penalty(target_length - 1))
+
+            # Return the sums and the target distributions
+            # sums: (batch_size,) target_dists: (batch_size, target_seq_len, target_vocab_size)
+            return mx.sym.Group([sums, target_dists]), data_names, label_names
+
+        if self.bucketing:
+            logger.info("Using bucketing. Default max_seq_len=%s", default_bucket_key)
+            self.module = mx.mod.BucketingModule(sym_gen=sym_gen,
+                                                 logger=logger,
+                                                 default_bucket_key=default_bucket_key,
+                                                 context=self.context)
+        else:
+            symbol, _, __ = sym_gen(default_bucket_key)
+            self.module = mx.mod.Module(symbol=symbol,
+                                        data_names=data_names,
+                                        label_names=label_names,
+                                        logger=logger,
+                                        context=self.context)
+
+        self.module.bind(data_shapes=provide_data,
+                         label_shapes=provide_label,
+                         for_training=False,
+                         force_rebind=False,
+                         grad_req=None)
```

Comment:
I think `grad_req` should be set to 'null', not None. This could save you memory :)
https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/module/executor_group.py#L182